### PR TITLE
Improve predicate push into RowNumberNode

### PR DESCRIPTION
**Pull request by kasiafi**

### CODE REVIEW

File: presto-main/src/main/java/io/prestosql/sql/planner/plan/RowNumberNode.java [hunk 1]
```diff
@@ -63,6 +63,7 @@ public RowNumberNode(
         checkArgument(!orderSensitive || partitionBy.isEmpty(), "unexpected partitioning in order sensitive node");
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/sql/planner/plan/RowNumberNode.java:63]: Make commit message more explicit: "Verify arguments to RowNumberNode constructor"

File: presto-main/src/main/java/io/prestosql/sql/planner/optimizations/WindowFilterPushDown.java [hunk 2]
```diff
@@ -169,8 +171,11 @@ public PlanNode visitFilter(FilterNode node, RewriteContext<Void> context)
                 OptionalInt upperBound = extractUpperBound(tupleDomain, rowNumberSymbol);
 
                 if (upperBound.isPresent()) {
+                    if (upperBound.getAsInt() <= 0) {
+                        return new ValuesNode(node.getId(), node.getOutputSymbols(), ImmutableList.of());
+                    }
                     source = mergeLimit(((RowNumberNode) source), upperBound.getAsInt());
-                    return rewriteFilterSource(node, source, rowNumberSymbol, upperBound.getAsInt());
+                    return rewriteFilterSource(node, source, rowNumberSymbol, ((RowNumberNode) source).getMaxRowCountPerPartition().get());
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/sql/planner/optimizations/WindowFilterPushDown.java:178]: I'm not sure I understand why this line changed. Can you explain?

Comment (kasiafi) [reply presto-main/src/main/java/io/prestosql/sql/planner/optimizations/WindowFilterPushDown.java:178]: After `mergeLimit()`, the source's `maxRowCountPerPartition` is either equal to `upperBound` or lower (the latter happens in the case when the source was already limited with a lower limit).

So, I decided to pass to `rewriteFilterSource()` the actual limit value of the source, which may be lower than `upperBound`. Consequently, the possibly lower bound value is tested against the original filter predicate, which improves the chance that we consider the predicate satisfied.

Example:
```
- Filter (row_num < 5 AND row_num BETWEEN 20 AND 30)
    - RowNumberNode (maxRowCountPerPartition == 3)

```
we get `upperBound = 30`, and then call `mergeLimit(30)` on RowNumberNode, which is ineffective, as the current limit is lower (`3`).

Next, we call `allRowNumberValuesInDomain()` to check if the established range for `row_num` symbol lies within the domain defined by Filter predicate.
I we pass `30`, as before the change, we get false.
If we pass the actual value `3`, we get true and can consider the predicate satisfied, and the Filter redundant.

**Comment by kasiafi** [comment]
Changed commit message.

**Pull request merged**